microsoft/PowerToys
# [PowerRename][ImageResizer] Tier1 Win11 Context menu

## Summary of the Pull Request
Add tier1 context menu items (Win11) for PowerRename and ImageResizer:
![image](https://user-images.githubusercontent.com/57057282/175318688-4a379c28-7e7f-4ffa-9aeb-366f15a25ed1.png)

Also, with tier1 context menu PowerRename can be started by right-clicking directory background (no need to select items):
![image](https://user-images.githubusercontent.com/57057282/175263197-eb2f5e4e-cba9-4b20-acb6-83b3f48617ca.png)


## PR Checklist

- [x] **Closes:** #13168


## Detailed Description of the Pull Request / Additional comments
Tier1 context menu are added by registering sparse packages for both PowerRename and ImageResizer. Packages are registered per user on module->enable() (conditionally if win11 and if not already registered). Packages are unregistered on PowerToys uninstall.

Also, minimum supported windows version is bumped to 10.0.19041.0 as sparse packaged registration API is added there. Also, end of support for previously used min version (10.0.18362.0) was 2022-05-10.

ImageResizer settings and trace files are extracted to ImageResizerLib project, to be able to consume those both from Ext and ContextMenu projects.

As CreateProcess can't run new proces from packaged app (sparse package here), PowerRename and ImageResuzer are being ran with explorer as parent process. Therefore, additional logic is added to use named pipes in that case (comparing to default stdin anonymous pipes when standard context menu is used).


## Validation Steps Performed
 - Build installer and install PowerToys on Win11
 - Sign sparse packages manually in order to be registered successfully
 - Enable PowerRename and ImageResizer modules (if not enabled)
 - Check from PowerShell that packages are registered e.g. `powershell -c get-appxpackage -name *imageresizer*`
 - If packages are registered and items are not visible in tier1 context menu, explorer needs to be restarted (known issue)
